### PR TITLE
barcode perf tests

### PR DESCRIPTION
**Pull request by MaximMilashchenko**
Added four perf case for barcode. Detection and decoding time check.

### CODE REVIEW

**Review by opencv-alalek — CHANGES_REQUESTED** [review]

**Comment by opencv-alalek** [comment]
Preliminary results on i7-12700K with 1/2/4/8 threads (no fixed frequency, with turbo boost enabled):

SIMD:

|Name of Test|1th|2th|4th|8th|
|---|:-:|:-:|:-:|:-:|
|decode::Perf_Barcode_multi::"4_barcodes.jpg"|2.500|2.177|2.033|1.989|
|decode::Perf_Barcode_single::"book.jpg"|0.415|0.389|0.390|0.376|
|decode::Perf_Barcode_single::"bottle_1.jpg"|0.266|0.221|0.198|0.221|
|decode::Perf_Barcode_single::"bottle_2.jpg"|0.712|0.699|0.732|0.727|
|detect::Perf_Barcode_multi::"4_barcodes.jpg"|7.527|5.939|5.486|4.888|
|detect::Perf_Barcode_single::"book.jpg"|1.679|1.660|1.670|1.685|
|detect::Perf_Barcode_single::"bottle_1.jpg"|3.243|2.687|2.447|2.326|
|detect::Perf_Barcode_single::"bottle_2.jpg"|1.974|1.985|1.975|2.045|

noSIMD / noIPP (`-DCV_DISABLE_OPTIMIZATION=ON`):

|Name of Test|1th|2th|4th|8th|
|---|:-:|:-:|:-:|:-:|
|decode::Perf_Barcode_multi::"4_barcodes.jpg"|4.615|3.477|3.162|3.060|
|decode::Perf_Barcode_single::"book.jpg"|0.650|0.584|0.586|0.579|
|decode::Perf_Barcode_single::"bottle_1.jpg"|0.536|0.406|0.341|0.295|
|decode::Perf_Barcode_single::"bottle_2.jpg"|0.975|0.881|0.872|0.871|
|detect::Perf_Barcode_multi::"4_barcodes.jpg"|9.504|7.636|6.563|6.274|
|detect::Perf_Barcode_single::"book.jpg"|2.580|2.500|2.493|2.570|
|detect::Perf_Barcode_single::"bottle_1.jpg"|4.240|3.750|3.507|3.270|
|detect::Perf_Barcode_single::"bottle_2.jpg"|2.964|2.896|2.839|2.852|

1 thread, "Generic C++" vs optimized code:

|Name of Test|generic|optimized|x-factor|
|---|:-:|:-:|:-:|
|decode::Perf_Barcode_multi::"4_barcodes.jpg"|4.615|2.500|1.85|
|decode::Perf_Barcode_single::"book.jpg"|0.650|0.415|1.57|
|decode::Perf_Barcode_single::"bottle_1.jpg"|0.536|0.266|2.02|
|decode::Perf_Barcode_single::"bottle_2.jpg"|0.975|0.712|1.37|
|detect::Perf_Barcode_multi::"4_barcodes.jpg"|9.504|7.527|1.26|
|detect::Perf_Barcode_single::"book.jpg"|2.580|1.679|1.54|
|detect::Perf_Barcode_single::"bottle_1.jpg"|4.240|3.243|1.31|
|detect::Perf_Barcode_single::"bottle_2.jpg"|2.964|1.974|1.50|

File: modules/barcode/perf/perf_barcode.cpp [hunk 1]
```diff
@@ -0,0 +1,85 @@
+#include "perf_precomp.hpp"
```
Comment (opencv-alalek) [inline modules/barcode/perf/perf_barcode.cpp:1]: License header

**Review by opencv-alalek — CHANGES_REQUESTED** [review]
Why is removed modules/barcode/test/test_barcode.cpp?

**Comment by MaximMilashchenko** [comment]
> Why is removed modules/barcode/test/test_barcode.cpp?

it is my fault, i'll correct it

File: modules/barcode/perf/perf_barcode.cpp (hunk 1 above)
Comment (MaximMilashchenko) [reply modules/barcode/perf/perf_barcode.cpp:1]: Do you mean the comment in the header?

**Comment by opencv-alalek** [comment]
@SinM9 please review this PR.

File: modules/barcode/perf/perf_barcode.cpp [hunk 2]
```diff
@@ -0,0 +1,89 @@
...
+    auto bardet = barcode::BarcodeDetector();
+    vector<cv::String> decoded_info;
+    vector<barcode::BarcodeType> decoded_type;
+
+    string image_path = findDataFile(root + name_current_image);
+    Mat src = imread(image_path);
+    ASSERT_FALSE(src.empty()) << "Can't read image: " << image_path;
+    vector< Point > corners;
+
+    cout << src.size << endl;
```
Comment (SinM9) [inline modules/barcode/perf/perf_barcode.cpp:79]: and here

File: modules/barcode/perf/perf_barcode.cpp [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+    vector< Point > corners;
+    string image_path = findDataFile(root + name_current_image);
+
+    Mat src = imread(image_path);
+    ASSERT_FALSE(src.empty()) << "Can't read image: " << image_path;
+
+    cout << src.size << endl;
+    bardet.detect(src, corners);
+
+    TEST_CYCLE() bardet.decode(src, corners, decoded_info, decoded_type);
```
Comment (SinM9) [inline modules/barcode/perf/perf_barcode.cpp:46]: TEST_CYCLE() ASSERT_TRUE(bardet.decode(src, corners, decoded_info, decoded_type)); ?

**Review by opencv-alalek — APPROVED** [review]
LGTM :+1:

File: modules/barcode/perf/perf_precomp.hpp [hunk 4]
```diff
@@ -0,0 +1,11 @@
+// This file is part of OpenCV project.
+// It is subject to the license terms in the LICENSE file found in the top-level directory
+// of this distribution and at http://opencv.org/license.html.
+
+#ifndef __OPENCV_PERF_PRECOMP_HPP___
+#define __OPENCV_PERF_PRECOMP_HPP__
```
Comment (opencv-alalek) [inline modules/barcode/perf/perf_precomp.hpp:6]: > `__OPENCV_PERF_PRECOMP_HPP__`

Both lines must be be the same. See GitHub Actions build error.

**Review by opencv-alalek — CHANGES_REQUESTED** [review]
Please squash commits into one before merge.

File: modules/wechat_qrcode/test/test_qrcode.cpp [hunk 5]
```diff
@@ -455,5 +455,16 @@ TEST_P(Objdetect_QRCode_Easy_Multi, regression) {
 std::string qrcode_model_path[] = {"", "dnn/wechat_2021-01"};
 INSTANTIATE_TEST_CASE_P(/**/, Objdetect_QRCode_Easy_Multi, testing::ValuesIn(qrcode_model_path));
 
+TEST(Objdetect_QRCode_bug, issue_3478) {
```
Comment (opencv-alalek) [inline modules/wechat_qrcode/test/test_qrcode.cpp:458]: Why do we have this here?

This is not your code.

File: modules/barcode/test/perf_barcode [hunk 6]
```diff
@@ -0,0 +1,85 @@
+#include "test_precomp.hpp"
```
[the remark this answers is not in the document]
Comment (opencv-alalek) [reply modules/barcode/test/perf_barcode:1]: We still have this invalid file in the patch.

**Review by opencv-alalek — CHANGES_REQUESTED** [review]

**Pull request merged**